Modules/_testcapimodule.c

```diff
@@ -2587,6 +2587,117 @@ get_datetime_fromtimestamp(PyObject* self, PyObject *args)
     return rv;
 }
 
+static PyObject *
+get_fields_from_date(PyObject *self, PyObject *args)
+{
+    PyObject *obj = NULL;
+    PyObject *rv;
+    int year, month, day;
+
+    if (!PyArg_ParseTuple(args, "O", &obj)) {
+        return NULL;
+    }
+
+    year = PyDateTime_GET_YEAR(obj);
```

Comment:
PyDateTime_GET_YEAR is documented to return an int... but does it really return a *signed* integer? Macro:

```
typedef struct
{
    _PyTZINFO_HEAD
    unsigned char data[_PyDateTime_DATE_DATASIZE];
} PyDateTime_Date;

#define PyDateTime_GET_YEAR(o)     ((((PyDateTime_Date*)o)->data[0] << 8) | \
                     ((PyDateTime_Date*)o)->data[1])
```

I don't know how the compiler promotes integer types when using unsigned char << 8.

Same question for PyDateTime_GET_MONTH and PyDateTime_GET_DAY which seem to return *unsigned char* (so unsigned integer) in practice.

@pganssle: What do you think of modifying these 3 macros to explicitly cast to int? By the way, I'm trying to move away from macros and use static inline functions instead. Would it make sense for these 3 macros?

Reply:
No valid `datetime` object will return a negative number for any of these values, so the difference between signed and unsigned integers in this case seems a bit academic. Are you suggesting that these tests would need to change if the macro were explicitly casting to int? Otherwise I think this is a separate issue unrelated to this PR.

> By the way, I'm trying to move away from macros and use static inline functions instead. Would it make sense for these 3 macros?

I'm not entirely sure what the criteria are for this *migrating* a macro to a function like this, but I would absolutely like to at least *expose* a function that does this. It will help with PyPy and other language compatibilities. Somewhere deep in my to do list is to explicitly document the capsule-based API and recommend it as the preferred way to interact with the `datetime` module in extension modules. That seems like a natural place to put accessor functions (though I admit I don't know how well that will play with the "static" and "inline" portions of your suggestion), with the added benefit of no backwards compatibility issues.

Reply:
static inline doesn't bring anything to PyPy and other languages: it's just to better define the return type and avoid macros uglyness. See https://bugs.python.org/issue35059 and https://gcc.gnu.org/onlinedocs/cpp/Macro-Pitfalls.html for example. See also § Convert macros to static inline functions: https://vstinner.github.io/split-include-directory-python38.html

> No valid datetime object will return a negative number for any of these values, so the difference between signed and unsigned integers in this case seems a bit academic. Are you suggesting that these tests would need to change if the macro were explicitly casting to int? Otherwise I think this is a separate issue unrelated to this PR.

It can be done in a separated PR. I was surprised by the usage of PyLong_FromSsize_t(): I don't expect a year to need size_t type, the datetime module is limited to year 9999 which fits into an int.

Maybe replace PyLong_FromSsize_t() with PyLong_FromLong()? :-)

Reply:
I guess changing from PyLong_FromSsize_t() to PyLong_FromLong() should only apply to the year. no? @vstinner

Reply:
You can use PyLong_FromLong() for all C variables of type "int". Which seems to be the case for all your variables, no?